### PR TITLE
Improve docs for `single_line_if_else_max_width` & add tests

### LINKED ISSUE #5496: Improve Documentation For `single_line_if_else_max_width` in `Configurations.md` + Add Tests
The [`single_line_if_else_max_width`](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#single_line_if_else_max_width)  is one of many configuration options that could use some improvements to it's documentation. It would be great to:

* Add code snippets to highlight the functionality this option provides
* Describe that the behavior of this option changes when `version=Two` is set.

Additionally, some tests for the options should be added to `tests/source/configs/` and `tests/target/configs/`. See the [Create test cases section of the Contributing guide](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#create-test-cases) for more details. It would be great to have test cases for both `version=One` and `version=Two`.

### PR DESCRIPTION
**Pull request by truls-p**
Closes #5496.

I am afraid I do not have much experience with ASTs and rustfmt in general, but I have tried to understand the implications of the `single_line_if_else_max_width` config option by stepping through the code and coming up with some examples + tests.

Please let me know if you have any comments or I have missed something significant that should be added to Configurations.md to further clarify this config option.

### CODE REVIEW

File: Configurations.md [hunk 1]
```diff
@@ -2379,14 +2379,48 @@ Don't reformat out of line modules
 
 ## `single_line_if_else_max_width`
 
-Maximum line length for single line if-else expressions. A value of `0` (zero) results in if-else expressions always being broken into multiple lines. Note this occurs when `use_small_heuristics` is set to `Off`.
+Maximum line length for single line if-else expressions. A value of `0` (zero) results in if-else expressions always being broken into multiple lines. Note this occurs when `use_small_heuristics` is set to `Off`. If one of the if-else blocks contain a statement, a comment, an attribute, or is empty, the if-else expression will be broken into multiple lines. When using `version = "Two"` the if-else expression at the end of a block will be formatted on a single line. The last if-else expression in the last example snippet of this section would be formatted on a single line with `version = "Two"`.
```
Comment (ytmimi) [inline Configurations.md:2382]: > When using `version = "Two"` the if-else expression at the end of a block will be formatted on a single line

Are you saying that this would only apply to the last if-else expression?  I don't think that's correct, but I could be wrong.

Comment (truls-p) [reply Configurations.md:2382]: Thanks for having a look! I agree that this behavior sounds strange and maybe my research and testing has come up short here, or maybe the proposed documentation is not clear enough. Let me try to explain what I see happening. When we try to use `single_line_if_else_max_width` in expr.rs `self.allow_single_line` needs to be `true` (see [here](https://github.com/truls-p/rustfmt/blob/improve-docs-if-else-max/src/expr.rs#L922)), which only happens when we have an `if` and expression type is == `ExprType::SubExpression` (see [here](https://github.com/truls-p/rustfmt/blob/improve-docs-if-else-max/src/expr.rs#L649)). Now, the `version = "Two"` is relevant [here](https://github.com/truls-p/rustfmt/blob/improve-docs-if-else-max/src/stmt.rs#L78) where we set `expr_type` to `ExprType::SubExpression`. The two tests I added, 70.rs and 70_version_two.rs tries to test this specific behavior.

Here is also another example to play around with. If you run rustfmt --check on this you will see we get a diff for both functions. However, if we change version to "Two" we only get a diff for the last `if` in `foo`.
Config options:
```
single_line_if_else_max_width = 80
version = "One"
```
main.rs:
```rust
fn bar() {}
fn foo() -> usize {
    let some_long_name = true;
    if some_long_name && some_long_name && some_long_name { bar() } else { bar() }
    return 1
}

fn baz() {
    let some_long_name = true;
    if some_long_name && some_long_name && some_long_name { bar() } else { bar() }
}

fn main() {
    let i = foo();
    baz();
    println!("{}", i);
}

```

Comment (ytmimi) [reply Configurations.md:2382]: Here are some other examples to consider:

What about something like this where the `if-else` is on the right hand side of an assignment expression?

```rust
fn foo() -> usize {
    let x = if some_long_name && some_long_name && some_long_name { bar() } else { bar() };
    let some_long_name = true;
    x
}
```

Also what about conditionally passing an argument to a function call or returning a value from a closure?
```rust
fn main() {
    let x = 0;
    f(if x == 0 { 0 } else { 10 })
    g(|x| if x == 0 { 0 } else { 10 })
}
```

Again, really appreciate your work on this, and helping us to nail down the description for this option. I think a little rewording is still needed, and maybe it's worth investigating how we set the `expr_type` argument. For example, when we call `rewrite` on a top level `ast::Expr` node we always pass `expr_type` as `ExprType::SubExpression`. See [`impl Rewrite for ast::Expr`](https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/expr.rs#L39-L43). The only place in the code base I could find where we don't hard code the `expr_type` argument to `format_expr` is when [it gets called from `format_stmt`](https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/stmt.rs#L111).

We happen to define our own borrowed version of a `ast::Stmt`:

https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/stmt.rs#L13-L16

And the rewrite method for that does follow the behavior that you outlined above where the last expression is set to a `ExprType::SubExpression` if `Version=Two` is also set

https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/stmt.rs#L76-L85

So maybe that's what explains the last expression behavior that you documented.


I also came across [this interesting comment](https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/expr.rs#L1038-L1041) in the code base that would be nice to have a test for.

> If one of the if-else blocks contain a statement, a comment, an attribute, or is empty, the if-else expression will be broken into multiple lines. 

It would also be great to have tests for each of these cases that you outlined!

Comment (truls-p) [reply Configurations.md]: Thanks for the detailed feedback so far!
These are good examples and I have added them to the documentation in the example section and mentioned them explicitly (the let, function call, and closure) in the description of the config option. I have also reworded the part about if-else expressions at the end of blocks. More on this below.

I have also added some more test cases for the nested if-else case (note how the inner if-else is on a single line when using `Version = "Two"` because it is the last expression in the block) and the comment about the if-else blocks needing to  be simple.

Regarding the `Version = "Two"` and how the `expr_type` is set I believe I now have a slightly better understanding about how this all works. My understanding is that as we walk the nodes everything starts out as a `Stmt`. This is why, for `Version = "One"`, when we look at an if-else at the end of a block we hit the following block of code and `expr_type` is set to `ExprType::Statement`. 

https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/stmt.rs#L76-L85

However, in the let, function call, and closure examples we set the `expr_type` for the if-else much further down the call stack. As an example, for the if-else inside the function call we call we hit `rewrite_cond` where the `expr_type` is set to `ExprType::SubExpression` by default.
https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/expr.rs#L610-L610

Then, crucially, `expr_type == ExprType::SubExpression` evaluates to true in `fn to_control_flow`.
https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/expr.rs#L645-L658

On the other hand, for the if-else at the end of a block we go straight from `impl<'a> Rewrite for Stmt<'a>` to `fn format_stmt` to `fn format_expr`, and finally to `fn control_flow`, where the `expr_type` is either `ExprType::Statement` or `ExprType::SubExpression` (set in `impl<'a> Rewrite for Stmt<'a>`) depending on whether we use `Version = "One"` or `Version = "Two"`.

I also dug up the original PR for the `impl<'a> Rewrite for Stmt<'a>`:
https://github.com/rust-lang/rustfmt/pull/3631
Note the tests added for #3614: https://github.com/rust-lang/rustfmt/pull/3631/files#diff-c331b0359e60165a708105a49c54a82dc01be48cd219a4c83f91ba0ba961dcb2

Please let me know your thoughts!

File: Configurations.md [hunk 2]
```diff
@@ -2381,12 +2381,62 @@ Don't reformat out of line modules
...
+        baz()
+    } else {
+        buzz()
+    };
+    if some_long_name && some_other_long_name {
+        1
+    } else {
+        2
+    }
+}
```
Comment (ytmimi) [inline Configurations.md:2420]: Would be nice to have some vertical space between the various cases so it's easier to tell what's going on, but besides that these examples look good 👍🏼. Also, the examples don't need to compile, they just need to parse, so it's actually not necessary to define the `some_long_name` and `some_other_long_name` variables. Just like `fun()`, `clousure()`, `baz()`, and `buzz()` aren't defined. Feel free to leave them in, just thought I'd point that out.

```suggestion
fn foo() -> usize {
    fun(if some_long_name && some_other_long_name {
        0
    } else {
        10
    });
    
    closure(|super_long_closure_variable| {
        if super_long_closure_variable == 0 {
            0
        } else {
            10
        }
    });
    
    let bar = if some_long_name && some_other_long_name {
        baz()
    } else {
        buzz()
    };
    
    if some_long_name && some_other_long_name {
        1
    } else {
        2
    }
}
```

File: Configurations.md [hunk 3]
```diff
@@ -2381,12 +2381,62 @@ Don't reformat out of line modules
...
+    let some_long_name = true;
+    let some_other_long_name = false;
+    fun(if some_long_name && some_other_long_name { 0 } else { 10 });
+    closure(|super_long_closure_variable| if super_long_closure_variable == 0 { 0 } else { 10 });
+    let bar = if some_long_name && some_other_long_name { baz() } else { buzz() };
+    if some_long_name && some_other_long_name {
+        1
+    } else {
+        2
+    }
```
Comment (ytmimi) [inline Configurations.md:2436]: Similar comment as above. I think it would be nice to have each example separated by a newline.

File: tests/source/issue-5496/70.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// rustfmt-version: One
+// rustfmt-single_line_if_else_max_width: 70
+fn foo() -> usize {
+    let some_long_name = true;
+    let some_other_long_name = false;
+    let bar = if some_long_name && some_other_long_name { baz() } else { buzz() };
+    if some_long_name && some_other_long_name { 1 } else { 2 }
+}
```
Comment (ytmimi) [inline tests/source/issue-5496/70.rs:8]: Not a big deal, but would be nice to have a newline at the end of the file. I believe most `source` tests are missing the trailing newline.

File: Configurations.md [hunk 5]
```diff
@@ -2381,12 +2381,62 @@ Don't reformat out of line modules
 
 Maximum line length for single line if-else expressions. A value of `0` (zero) results in if-else expressions always being broken into multiple lines. Note this occurs when `use_small_heuristics` is set to `Off`.
 
+Common if-else expressions that may be formatted on a single line are expressions on the right hand side of the equals (`=`) sign in a let statement, expressions inside a function call, and expressions inside a closure. If one of the if-else blocks contain a statement, a comment, an attribute, or is empty, the if-else expression will be broken into multiple lines. For backwards compatability reasons, an if-else expression at the end of a block is treated as a statement when using `version = "One"` and may not be formatted on a single line. When using `version = "Two"` an if-else expression at the end of a block may be formatted on a single line if it is a simple if-else expression as described earlier. As an example, the last if-else expression in the last snippet of this section would be formatted on a single line with `version = "Two"`.
```
Comment (ytmimi) [inline Configurations.md:2384]: This is a much more descriptive than what we used to have, thank you! Here are some suggestions on how to reformat it to make it easier for readers to understand since it's currently all just one big line of text:

```suggestion
Common if-else expressions that may be formatted on a single line are:
* Expressions on the right hand side of an equals (`=`) sign. e.g. `let x = if y == 1 { 0 } else { 1 };`
* Conditional expression in function calls `f(if x== 0 { 10 } else { 0 })`
* Single expression closures. e.g. `let x = |y| if y == 0 { 0 } else { 10 };` 

In the following scenarios the if-else expression will be formatted over multiple lines.
* Empty if-else blocks.
* An outer if-else expression when if-else expression are nested.
* if-else blocks that contain statements e.g. `let x = a;`, comments,  or attributes `let _ = if true { #[must_use] 1 } else { 2 };`.

When `version=One` is set (the default) if-else expression at the end of a block are treated as statements and are not formatted on a single line. When using `version=Two` the if-else expression may be formatted on a single line if it meets the criteria described above.
```

**Review by ytmimi — COMMENTED** [review]
Again, thanks for your work on this. Take a look at my suggestions and let me know if you have any questions.

File: tests/source/issue-5496/70.rs (hunk 4 above)
Comment (truls-p) [reply tests/source/issue-5496/70.rs]: Thanks, done!

File: Configurations.md (hunk 3 above)
Comment (truls-p) [reply Configurations.md]: I agree, done!

**Comment by truls-p** [comment]
Thanks again for all the help, @ytmimi ! I think I have now incorporated all the requested changes.

**Comment by ytmimi** [comment]
Great! I'll look to do a follow up after work

**Review by ytmimi — APPROVED** [review]
Thanks for all the help with updating the docs and your first contribution to rustfmt 🎉. I think we're good to go, but will hold off on merging to let @calebcartwright take a look!

File: Configurations.md [hunk 6]
```diff
@@ -2381,12 +2381,74 @@ Don't reformat out of line modules
...
+* Conditional expression in function calls. e.g. `f(if x== 0 { 10 } else { 0 })`
+* Single expression closures. e.g. `let x = |y| if y == 0 { 0 } else { 10 };` 
+
+In the following scenarios the if-else expression will be formatted over multiple lines.
+* Empty if-else blocks.
+* An outer if-else expression when if-else expression are nested.
+* If-else blocks that contain statements (e.g. `let x = a;`), comments, or attributes (e.g. `let _ = if true { #[must_use] 1 } else { 2 };`).
+
+When `version=One` is set (the default) an if-else expression at the end of a block is treated as a statement and is not formatted on a single line. When using `version=Two` the if-else expression may be formatted on a single line if it meets the criteria described above.
+
```
Comment (calebcartwright) [inline Configurations.md:2395]: I have some reservations about any attempts to reference, or perhaps to some readers, codify formatting behavior here. The Style Guide is what articulates the rules for the formatting behavior, and the exact prescriptions for when `if-else` can/can't be single lined can be found [here](https://github.com/rust-lang/rust/blob/master/src/doc/style-guide/src/expressions.md#single-line-if-else).

This option exists as the mechanism to control the value used as the "small" threshold as described in the Guide, and I don't think we should have much more than that here.

Basically, the Config docs, and any particular option within them, isn't really the right place to try to explain the broader formatting rules for a given syntax construct

**Review by calebcartwright — CHANGES_REQUESTED** [review]
Thanks for the contribution! I imagine the tests are probably useful in their own right, but I think this is the wrong direction to take with the documentation updates.

I've shared more details below, and will probably have some related thoughts to share on the associated issue
